### DISCUSSION TITLE
I don't understand how I am supposed to use it

### DISCUSSION THREAD

**Discussion by ev-agelos** [mjpieters/aiolimiter#72] [Q&A] [answered]
The API I am trying against has a limit of 250 requests per minute. Whatever I try I get 429 responses, it doesnt seep to "sleep" or "slow down" after 250 requests
```python
async def main():
    limiter = AsyncLimiter(250, 60)
    async with client as cl:  # httpx async client
        # for i in range(400):
        #     async with limiter:
        async with limiter:
            for i in range(400):
                await cl.get(url)

asyncio.run(main())
```
Furthermore, the code I am trying to make use of your library has this form:
```python
async def get_name():
    # api request
    ...

async def get_last_name():
    # another api request
    ...

async def get_user_data():
    name = await get_name()
    last_name = await get_last_name()
    return name, last_name

async def main():
    coros = [get_user_data() for user in many_users()]
    for result in asyncio.gather(*coros):
        # handle result
```
Is it possible somehow to apply the limiter in such structure?

**Comment by mjpieters** [comment, marked as answer]
Your error lies here:

> ```python
>         async with limiter:
>             for i in range(400):
>                 await cl.get(url)
> ```

The **whole** block under `async with limiter:` is limited to 256 executions per 60 seconds, but you loop inside that block. Put the loop **outside** of the limiter:

```python
        for i in range(400):
            async with limiter:
                await cl.get(url)
```

Now execution passes from `for` to the `async with limiter:` block, which can then control how fast `await cl.get(url)` is called.

> Is it possible somehow to apply the limiter in such structure?

Create a limiter in `main()` and pass it into `get_user_data()` so it is shared between coroutines:

```python
async def get_user_data(limiter):
    with limiter:
        name = await get_name()
        last_name = await get_last_name()
    return name, last_name

async def main():
    limiter = AsyncLimiter(250, 60)
    coros = [get_user_data(limiter) for user in many_users()]
    for result in asyncio.gather(*coros):
        # handle result
```

**Comment by ev-agelos** [comment]
Thanks for your response @mjpieters. I tried what you suggest as well, if you see in my post I have your suggestion as comment. I am going to give it another try again.
~Hmm i think it is working. I see that the requests start to spread in time. Unfortunately I am getting 429 which means I need to figure out how the server calculates the time.~
I think it doesnt work as expected, so the first request happened at `12:35:34` and the 250ish at `12:35:43` and it continued firing at `12:35:44`, `12:35:45` and so on. The minute hasnt passed yet and it continued firing requests, is that expected? I thought it would wait for the minute to pass and then continue firing.

**Comment by mjpieters** [comment]
The limiter will allow the first 250 requests through at once, see the [*Bursting* section](https://aiolimiter.readthedocs.io/en/latest/#bursting) of the documentation; requests are not automatically spread out over the time period.

If you are still not seeing a limit on the number of requests in the time period you set, you are not yet correctly using the limiter. Try the example in the 'bursting' section I linked to to see the limiter in operation.

**Comment by Abdelgha-4** [comment]
I do understand the behavior of the limiter, but if an API has a rate limit of 4 requests per 8 seconds, the example from the [Bursting section](https://aiolimiter.readthedocs.io/en/latest/#bursting) will trigger a 429 response because 6 requests are done in 8 seconds, not just 4.
